lib/teleterm/services/desktop/directorysharing.go

```diff
@@ -0,0 +1,313 @@
+	resolved, err := filepath.EvalSymlinks(full)
+	if err != nil {
+		// EvalSymlinks returns an error if the target file does not exist.
+		// In that case, attempt to resolve the symlinks of the parent directory instead.
+		if errors.Is(err, fs.ErrNotExist) {
+			parent := filepath.Dir(full)
+			resolvedParent, perr := filepath.EvalSymlinks(parent)
+			if perr != nil {
+				return "", trace.Wrap(perr)
+			}
+
+			// Reconstruct the full path by joining the resolved parent with the original file name.
+			resolved = filepath.Join(resolvedParent, filepath.Base(full))
+		} else {
+			return "", trace.Wrap(err)
+		}
+	}
+	if !isSubPath(d.basePath, resolved) {
+		return "", trace.BadParameter("path escapes from parent")
+	}
+	return resolved, nil
+}
+
+func isSubPath(parent, child string) bool {
+	return child == parent || strings.HasPrefix(child, parent+string(filepath.Separator))
+}
+
+// Stat retrieves metadata about a file or directory at the given path.
+func (d *DirectoryAccess) Stat(relativePath string) (*FileOrDirInfo, error) {
+	path, err := d.getSafePath(relativePath)
+	if err != nil {
+		return nil, trace.Wrap(err)
+	}
+
+	stat, err := os.Stat(path)
+	if err != nil {
+		return nil, trace.Wrap(err)
+	}
+
+	info, err := d.readFileOrDirInfo(relativePath, stat)
+	return info, trace.Wrap(err)
+}
+
+// ReadDir lists files and directories within the given directory path, skips symlinks.
+func (d *DirectoryAccess) ReadDir(relativePath string) ([]*FileOrDirInfo, error) {
+	path, err := d.getSafePath(relativePath)
+	if err != nil {
+		return nil, trace.Wrap(err)
+	}
+
+	entries, err := os.ReadDir(path)
+	if err != nil {
+		return nil, trace.Wrap(err)
+	}
+
+	var results []*FileOrDirInfo
+	for _, entry := range entries {
+		fileInfo, err := entry.Info()
+		if err != nil {
+			return nil, trace.Wrap(err)
+		}
+
+		// Skip symlinks, we can't present them properly in the remote machine.
+		if fileInfo.Mode().Type()&os.ModeSymlink != 0 {
+			continue
+		}
+
+		entryRelativePath := filepath.Join(relativePath, fileInfo.Name())
+		fileOrDir, err := d.readFileOrDirInfo(entryRelativePath, fileInfo)
+		if err != nil {
+			return nil, trace.Wrap(err)
+		}
+
+		results = append(results, fileOrDir)
+	}
+
+	return results, nil
+}
+
+// Read reads a slice of a file into buf. Returns the number of read bytes.
+func (d *DirectoryAccess) Read(relativePath string, offset int64, buf []byte) (n int, err error) {
+	path, err := d.getSafePath(relativePath)
+	if err != nil {
+		return 0, trace.Wrap(err)
+	}
+
+	file, err := os.Open(path)
+	if err != nil {
+		return 0, trace.Wrap(err)
+	}
+	defer func() {
+		closeErr := file.Close()
+		if err == nil {
+			// Only update err if no previous error occurred.
+			err = trace.Wrap(closeErr)
+		}
+	}()
+
+	_, err = file.Seek(offset, io.SeekStart)
+	if err != nil {
+		return 0, trace.Wrap(err)
+	}
+
+	n, err = file.Read(buf)
+	if err != nil && !errors.Is(err, io.EOF) {
+		return 0, trace.Wrap(err)
+	}
+	return n, trace.Wrap(err)
```

Comment:
```suggestion
	return n, err
```

No need to wrap, right? Looks like we've already wrapped every case of early return as well as the closeErr.